apps/emqx_gcp_device/src/emqx_gcp_device_authn.erl

```diff
@@ -0,0 +1,233 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_gcp_device_authn).
+
+-include_lib("emqx_authn/include/emqx_authn.hrl").
+-include_lib("emqx/include/logger.hrl").
+-include_lib("hocon/include/hoconsc.hrl").
+-include_lib("jose/include/jose_jwt.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-behaviour(hocon_schema).
+
+-export([
+    namespace/0,
+    tags/0,
+    roots/0,
+    fields/1,
+    desc/1
+]).
+
+-export([
+    refs/0,
+    union_member_selector/1,
+    create/2,
+    update/2,
+    authenticate/2,
+    destroy/1
+]).
+
+%%------------------------------------------------------------------------------
+%% Hocon Schema
+%%------------------------------------------------------------------------------
+
+namespace() -> "authn".
+
+tags() ->
+    [<<"Authentication">>].
+
+%% used for config check when the schema module is resolved
+roots() ->
+    [
+        {?CONF_NS,
+            hoconsc:mk(
+                hoconsc:union(fun ?MODULE:union_member_selector/1),
+                #{}
+            )}
+    ].
+
+fields(gcp_device) ->
+    common_fields().
+
+desc(gcp_device) ->
+    ?DESC(emqx_gcp_device_api, gcp_device);
+desc(_) ->
+    undefined.
+
+%%------------------------------------------------------------------------------
+%% APIs
+%%------------------------------------------------------------------------------
+
+refs() ->
+    [
+        hoconsc:ref(?MODULE, gcp_device)
+    ].
+
+union_member_selector(all_union_members) ->
+    refs();
+union_member_selector({value, Value}) ->
+    refs(Value).
+
+create(_AuthenticatorID, _Config) ->
+    {ok, #{}}.
+
+update(
+    _Config,
+    State
+) ->
+    {ok, State}.
+
+authenticate(#{auth_method := _}, _) ->
+    ignore;
+authenticate(Credential, _State) ->
+    check(Credential).
+
+destroy(_State) ->
+    emqx_gcp_device:clear_table(),
+    ok.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+common_fields() ->
+    [
+        {mechanism, emqx_authn_schema:mechanism('gcp_device')}
+    ] ++ emqx_authn_schema:common_fields().
+
+refs(#{<<"mechanism">> := <<"gcp_device">>}) ->
+    [hoconsc:ref(?MODULE, gcp_device)];
+refs(_) ->
+    throw(#{
+        field_name => mechanism,
+        expected => "gcp_device"
+    }).
```

Comment:
Q: will there be other mechanisms for GCP authn?

Since this module is specific to GCP and there's a single option for it, it seems that using an union selector might be unnecessary.